Add EVEOnlineV2 provider

Dusty-Meg opened the pull request:
Add provider for the V2 routes of the EVE Online SSO.

The V1 (current) routes are still valid, but they will eventually be phased out so having the V2 routes available will be useful. 
The V2 login pattern is the recommended route from the developers.

Dusty-Meg committed:
Add EVEOnlineV2 provider

kevinchalet commented:
Thanks for your contribution.

I don't think we should introduce a new provider. If we did that each time a service introduces a new API, the number of packages we have to maintain would literally explode.

Instead, can you please update the existing provider to use the V2 endpoints?

martincostello commented:
Yep same comments as Kévin - just migrate the internals of the existing provider.

We've already done the same for a number of providers in previous releases, such as Instagram (#461), LinkedIn (#298) and Yahoo (#393).

kevinchalet commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,81 @@ namespace AspNet.Security.OAuth.EVEOnline
-            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
-            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+            // JwtSecurityToken decodedToken = new JwtSecurityToken(tokens.AccessToken);
+            var tokenClaims = ExtractClaimsFromToken(tokens.AccessToken);
```
Whattttttttttt? Reading and assuming access tokens have a specific format - JWT here - in OAuth 2.0 client libs is a horrible practice. Is this really the pattern they are recommending now?

Dusty-Meg replied:
I am not sure what you mean here?

In terms of the AccessToken being a JWT that is what the devs are saying it is.
Documented [https://docs.esi.evetech.net/docs/sso/web_based_sso_flow.html](https://docs.esi.evetech.net/docs/sso/web_based_sso_flow.html)

kevinchalet replied:
Access tokens are opaque for the clients, that are not supposed to read them: they are only meant to be read by the resource APIs they target. Doing that is a serious protocol violation (that's why we have a different token called `id_token` in OpenID Connect).

Do we know when the v1 endpoints will be deprecated? I'd prefer if we waited a bit before adopting this insane non-standard approach.

Dusty-Meg replied:
Ok, thats fair enough.

We have no set day, all we have so far been told is soon, which in terms of the EVE Online developers could be anywhere from tomorrow to some point after the heat death of the universe.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationConstants.cs:
```diff
@@ -1,19 +0,0 @@
-    /// <summary>
-    /// Contains constants specific to the <see cref="EVEOnlineAuthenticationHandler"/>.
-    /// </summary>
-    public static class EVEOnlineAuthenticationConstants
```
Please restore this file and mark it as `[Obsolete]` with an appropriate message instead. Otherwise this is a binary breaking change.

Dusty-Meg replied:
I have re-added this and also added the code to add the scopes to the claim which I missed.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -48,11 +48,6 @@ public static class Tranquility
-            /// <summary>
-            /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
-            /// </summary>
-            public const string UserInformationEndpoint = "https://login.eveonline.com/oauth/verify";
```
Restore and mark as `[Obsolete]` with an appropriate message.

Dusty-Meg committed:
Correct failing tests in EVEOnline provider

Dusty-Meg commented:
I had left them separate so that people could still use the V1 endpoints, but guess that is what Nuget Packages are for.
Updated the pull request.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,81 @@ namespace AspNet.Security.OAuth.EVEOnline
+            // Unix timestamp is seconds past epoch
+            DateTime dateTime = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
+            dateTime = dateTime.AddSeconds(double.Parse(unixTimeStamp, CultureInfo.InvariantCulture));
+            return dateTime.ToString("yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'fff'Z'", CultureInfo.InvariantCulture);
```
I think the `"o"` format string does this for you: https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#the-round-trip-o-o-format-specifier

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,81 @@ namespace AspNet.Security.OAuth.EVEOnline
+            }
+        }
+
+        private static string UnixTimeStampToDateTime(string unixTimeStamp)
```
This can all be replaced by using either `DateTimeOffset.FromUnixTimeMilliseconds()` or `DateTimeOffset.FromUnixTimeSeconds()`.

Dusty-Meg replied:
Thanks for this, had not seen these before.

Dusty-Meg committed:
Add scopes back for EVEOnline provider

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -42,16 +42,17 @@ public static class Tranquility
             /// <summary>
             /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
             /// </summary>
+            [Obsolete("UserInformationEndpoint is no longer required for V2, info now in JWT token.")]
```
```suggestion
            [Obsolete("This value is no longer required for V2, information is now stored in the JWT.")]
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -63,16 +64,17 @@ public static class Singularity
             /// <summary>
             /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
             /// </summary>
+            [Obsolete("UserInformationEndpoint is no longer required for V2, info now in JWT token.")]
```
```suggestion
            [Obsolete("This value is no longer required for V2, information is now stored in the JWT.")]
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
-            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
-            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
-            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+            // JwtSecurityToken decodedToken = new JwtSecurityToken(tokens.AccessToken);
```
Left over commented-out code.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+        private static string UnixTimeStampToDateTime(string unixTimeStamp)
+        {
+            DateTimeOffset offset = DateTimeOffset.FromUnixTimeSeconds(long.Parse(unixTimeStamp, CultureInfo.InvariantCulture));
+            return offset.ToString("o");
```
Use the overload that takes `CultureInfo.InvariantCulture`.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+
+        private static string UnixTimeStampToDateTime(string unixTimeStamp)
+        {
+            DateTimeOffset offset = DateTimeOffset.FromUnixTimeSeconds(long.Parse(unixTimeStamp, CultureInfo.InvariantCulture));
```
Use `long.TryParse()` so that you can throw a more useful exception message if the parsing fails.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+                var claims = new List<Claim>(securityToken.Claims)
+                {
+                    new Claim(ClaimTypes.NameIdentifier, securityToken.Subject.Replace("CHARACTER:EVE:", string.Empty, StringComparison.OrdinalIgnoreCase), ClaimValueTypes.String, ClaimsIssuer),
+                    new Claim(ClaimTypes.GivenName, securityToken.Claims.First(x => x.Type.Equals("name", StringComparison.OrdinalIgnoreCase)).Value, ClaimValueTypes.String, ClaimsIssuer),
```
Don't use LINQ's `First()` method because if the value isn't found you'll get a very unhelpful exception message to try and debug.

martincostello replied:
Use the static `string.Equals(string, string, StringComparison)` method rather than the instance `Equals()` method as the former is safe against null values.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+
+                var claims = new List<Claim>(securityToken.Claims)
+                {
+                    new Claim(ClaimTypes.NameIdentifier, securityToken.Subject.Replace("CHARACTER:EVE:", string.Empty, StringComparison.OrdinalIgnoreCase), ClaimValueTypes.String, ClaimsIssuer),
```
I would do this outside a collection initialiser as you're doing non-trivial logic. If an exception was thrown by any of these four lines you wouldn't get a useful line number from the stack trace as to which one had failed.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+                var claims = new List<Claim>(securityToken.Claims)
+                {
+                    new Claim(ClaimTypes.NameIdentifier, securityToken.Subject.Replace("CHARACTER:EVE:", string.Empty, StringComparison.OrdinalIgnoreCase), ClaimValueTypes.String, ClaimsIssuer),
+                    new Claim(ClaimTypes.GivenName, securityToken.Claims.First(x => x.Type.Equals("name", StringComparison.OrdinalIgnoreCase)).Value, ClaimValueTypes.String, ClaimsIssuer),
```
You're using `name` as the Name and GivenName - is that correct? What if the name is a full name like _John Smith_?

Dusty-Meg replied:
The setting of GivenName is redundant and not needed as Name is being set, will drop that line out.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -20,44 +22,90 @@ namespace AspNet.Security.OAuth.EVEOnline
+                {
+                    claims.RemoveAll(x => scopes.Contains(x));
+
+                    claims = claims.Append(new Claim(EVEOnlineAuthenticationConstants.Claims.Scopes, string.Join(" ", scopes.Select(x => x.Value)), ClaimValueTypes.String, ClaimsIssuer)).ToList();
```
You already have a list here so you can just add the one value to it:

```suggestion
                    claims.Add(new Claim(EVEOnlineAuthenticationConstants.Claims.Scopes, string.Join(" ", scopes.Select(x => x.Value)), ClaimValueTypes.String, ClaimsIssuer));
```

Dusty-Meg committed:
Add error handling where missing in EVEOnline provider

Dusty-Meg commented:
Updated with your changes.

I would have added some tests around the exceptions but, cannot see how to change which item from the bundle.json a test uses.
Changed it around manually for the tests, but returned to normal before pushing.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -102,10 +103,27 @@ protected virtual IEnumerable<Claim> ExtractClaimsFromToken([NotNull] string tok
+
+            if (extractedClaim == null)
+            {
+                throw new MissingFieldException($"The claim '{claim}' missing from the JWT token from EVEOnline.");
```
```suggestion
                throw new InvalidOperationException($"The claim '{claim}' is missing from the EVEOnline JWT.");
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -102,10 +103,27 @@ protected virtual IEnumerable<Claim> ExtractClaimsFromToken([NotNull] string tok
-            return offset.ToString("o");
+            if (!long.TryParse(unixTimeStamp, NumberStyles.Any, CultureInfo.InvariantCulture, out long unixTime))
+            {
+                throw new InvalidCastException("The value of the claim 'exp' is not in the format long.");
```
```suggestion
                throw new InvalidOperationException("The value '{unixTimeStamp}' of the 'exp' claim is not a valid 64-bit integer.");
```

Dusty-Meg committed:
Alter exceptions from EVEOnline provider to InvalidOperationException

martincostello commented:
Just one outstanding PR comment each from myself and Kévin.

Dusty-Meg committed:
Remove redundant setting of GivenName claim for EVEOnline provider

martincostello commented:
If you would like to continue with this contribution please resolve the remaining comments and rebase to target the latest changes in the `dev` branch for .NET 5.0.

Otherwise we'll close this PR.

Dusty-Meg committed:
Rebased

Dusty-Meg commented:
> If you would like to continue with this contribution please resolve the remaining comments and rebase to target the latest changes in the `dev` branch for .NET 5.0.
> 
> Otherwise we'll close this PR.

Rebased, the remaining comment is not something I can see fixing, it is the way that CCP (EVE Onlines dev's) set out, and we still have no date set for the deprecation of the V1 (The version that is currently in the repo).

martincostello commented:
Thanks.

What do you want to do about the remaining comment @kevinchalet?

kevinchalet commented:
@kevinchalet giving that targeting the V2 endpoints requires doing horrible non-standard things, I'd be tempted to wait until the V1 endpoints are no longer supported.

Dusty-Meg commented:
Hi,
Update to this now we have a date,
The V1 versions will be deprecated on the 1st November 2021.
[https://developers.eveonline.com/blog/article/sso-endpoint-deprecations-2](https://developers.eveonline.com/blog/article/sso-endpoint-deprecations-2)

martincostello commented:
OK thanks - in that case I'll take a look at that doc at some point this week and re-review this PR and we can take it from there.

martincostello commented:
Sorry this fell through the cracks and I forgot about it - based on the dates you provided, does this still need addressing now that V1 is deprecated as of now?

This PR will need rebasing (or a new one opening in its place).

Dusty-Meg committed:
Merge branch 'dev' into dev

Dusty-Meg commented:
Yes the V1 version is now deprecated.

Have rebased.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationExtensions.cs:
```diff
@@ -4,7 +4,12 @@
  * for more information concerning the license and the contributors participating to this project.
  */
 
+using System;
+using System.IdentityModel.Tokens.Jwt;
 using AspNet.Security.OAuth.EVEOnline;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.DependencyInjection.Extensions;
```
I feel some of these are probably redundant now we're using implicit using statements in C# 10.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -65,11 +65,45 @@ public static class Singularity
         /// <summary>
         /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
         /// </summary>
-        public static readonly string TokenEndpoint = "https://sisilogin.testeveonline.com/oauth/token";
```
I think this rebase has gone awry.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -4,9 +4,17 @@
  * for more information concerning the license and the contributors participating to this project.
  */
 
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
 using System.Net.Http.Headers;
 using System.Security.Claims;
 using System.Text.Encodings.Web;
+using System.Threading.Tasks;
```
Again, the majority of these will now be redundant with implicit namespaces.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -23,33 +31,110 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
         : base(options, logger, encoder, clock)
     {
```
What happens with this constructor if there's no way to get a `JwtSecurityTokenHandler`? Will this additional constructor create DI ambiguity?

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationOptions.cs:
```diff
@@ -37,20 +29,21 @@ public EVEOnlineAuthenticationServer Server
+                        break;
+
+                    case EVEOnlineAuthenticationServer.Singularity:
+                        AuthorizationEndpoint = EVEOnlineAuthenticationDefaults.Singularity.AuthorizationEndpoint;
+                        TokenEndpoint = EVEOnlineAuthenticationDefaults.Singularity.TokenEndpoint;
+                        break;
+
+                    default:
+                        throw new ArgumentException($"Server '{value}' is unsupported.", nameof(value));
+                }
```
Again, looks like an incorrect rebase.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/EVEOnline/EVEOnlineTests.cs:
```diff
@@ -1,4 +1,4 @@
-﻿/*
+ /*
```
```suggestion
/*
```

Dusty-Meg committed:
Fixed broken rebase

Dusty-Meg commented:
Sorry about that, thought I could do the rebase via the online editor, seems not.

Should all be working now and adjusted to use Microsoft.IdentityModel.JsonWebTokens over System.IdentityModel.Tokens.Jwt inline with the apple provider.

martincostello commented:
Thanks - I'll take a look at this tomorrow 👍

Dusty-Meg commented:
I have no idea what has caused the builds to fail. Tried to work it out but out of ideas.

martincostello commented:
I know what that is - I'll fix it tomorrow, then a merge from dev should resolve it.

martincostello commented:
Had a quick chance to sort this out today, should be fixed now if you pull the latest commit from `dev`.

Dusty-Meg committed:
Merge branch 'aspnet-contrib:dev' into dev

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -31,45 +31,41 @@ public static class EVEOnlineAuthenticationDefaults
     /// </summary>
     public static readonly string CallbackPath = "/signin-eveonline";
 
-    /// <summary>
-    /// Default values for the Tranquility (live) server.
-    /// </summary>
```
Any particular reason you deleted this?

Dusty-Meg replied:
Believe this was most likely just being overzealous with the rebase fixing.
Will add them back.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -31,45 +31,41 @@ public static class EVEOnlineAuthenticationDefaults
         public static readonly string UserInformationEndpoint = "https://login.eveonline.com/oauth/verify";
     }
 
-    /// <summary>
-    /// Default values for the Singularity (test) server.
-    /// </summary>
```
+1

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -31,45 +31,41 @@ public static class EVEOnlineAuthenticationDefaults
         /// <summary>
         /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
         /// </summary>
+        [Obsolete("This value is no longer required for V2, information is now stored in the JWT.")]
```
Using "V2" in the message might confuse people unfamiliar with the internals of the provider.

```suggestion
        [Obsolete("This endpoint is no longer used by the EVEOnline provider.")]
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationDefaults.cs:
```diff
@@ -31,45 +31,41 @@ public static class EVEOnlineAuthenticationDefaults
         /// <summary>
         /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
         /// </summary>
+        [Obsolete("This value is no longer required for V2, information is now stored in the JWT.")]
```
```suggestion
        [Obsolete("This endpoint is no longer used by the EVEOnline provider.")]
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
-        [NotNull] ClaimsIdentity identity,
-        [NotNull] AuthenticationProperties properties,
-        [NotNull] OAuthTokenResponse tokens)
+          [NotNull] ClaimsIdentity identity,
+          [NotNull] AuthenticationProperties properties,
+          [NotNull] OAuthTokenResponse tokens)
```
```suggestion
        [NotNull] ClaimsIdentity identity,
        [NotNull] AuthenticationProperties properties,
        [NotNull] OAuthTokenResponse tokens)
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
         {
-            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
-            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+            throw new InvalidOperationException("No Access token was returned in the OAuth token.");
```
```suggestion
            throw new InvalidOperationException("No access token was returned in the OAuth token.");
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
+
+            if (scopes.Any())
+            {
+                claims.RemoveAll(x => scopes.Contains(x));
```
What's the rationale behind removing these `scp` claims? Why not just leave them as-is and just use them to create the `urn:eveonline:scopes` claim?

Dusty-Meg replied:
I have no recollection on why I put that line of code in, have removed and still working via the sample app.
Will update with line removed.

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
+
+            var scopes = claims.Where(x => string.Equals(x.Type, "scp", StringComparison.OrdinalIgnoreCase)).ToList();
+
+            if (scopes.Any())
```
```suggestion
            if (scopes.Count > 0)
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
+            {
+                claims.RemoveAll(x => scopes.Contains(x));
+
+                claims.Add(new Claim(EVEOnlineAuthenticationConstants.Claims.Scopes, string.Join(" ", scopes.Select(x => x.Value)), ClaimValueTypes.String, ClaimsIssuer));
```
```suggestion
                claims.Add(new Claim(EVEOnlineAuthenticationConstants.Claims.Scopes, string.Join(' ', scopes.Select(x => x.Value)), ClaimValueTypes.String, ClaimsIssuer));
```

martincostello commented on src/AspNet.Security.OAuth.EVEOnline/EVEOnlineAuthenticationHandler.cs:
```diff
@@ -25,31 +32,94 @@ public partial class EVEOnlineAuthenticationHandler : OAuthHandler<EVEOnlineAuth
+
+    private static string UnixTimeStampToDateTime(string unixTimeStamp)
+    {
+        if (!long.TryParse(unixTimeStamp, NumberStyles.Any, CultureInfo.InvariantCulture, out long unixTime))
```
```suggestion
        if (!long.TryParse(unixTimeStamp, NumberStyles.Integer, CultureInfo.InvariantCulture, out long unixTime))
```

Dusty-Meg committed:
Pull request edits for EVEOnline provider

Dusty-Meg committed:
Merge branch 'dev' of github.com:Dusty-Meg/AspNet.Security.OAuth.Providers into dev

martincostello reviewed (approved):
Looks good to me, but will defer to @kevinchalet for final approval as he had the initial reservations when the PR was first opened.

kevinchalet commented:
I guess EVE Online won't change their implementation at this point, so... yeah, let's do it 🤷🏻

martincostello commented:
Thanks again @Dusty-Meg - if you could just validate the changes in your app using the [`6.0.1-preview.21601.38`](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.EVEOnline/6.0.1-preview.21601.38) package from our prerelease feed, if everything's good then I'll push this out to NuGet.org.

martincostello commented:
Hey there @Dusty-Meg - have you had any chance to validate your changes?

Dusty-Meg commented:
Sorry about the delay, (new job)
I have just confirmed the changes are working using the preview version.

Thank you for being so patient, with this whole change.

martincostello commented:
No problem!

Thanks, I'll push the new 6.0.1 packages up to NuGet.org at some point this weekend.

martincostello commented:
This change has now been published to [NuGet.org](https://www.nuget.org/packages/AspNet.Security.OAuth.EVEOnline/6.0.1).

Thanks again 🥳
